conans/client/cmd/uploader.py

```diff
@@ -111,18 +111,32 @@ def _upload(self, conan_file, conan_ref, packages_ids, retry, retry_wait,
                                      "no packages can be uploaded")
             total = len(packages_ids)
             for index, package_id in enumerate(packages_ids):
-                ret_upload_package = self._upload_package(PackageReference(conan_ref, package_id),
+                pref = PackageReference(conan_ref, package_id)
+                cur_package_remote = self._registry.prefs.get(pref)
+
+                if remote_name:  # If remote_name is given, use it
+                    p_remote = self._registry.remotes.get(remote_name)
+                else:
+                    p_remote = cur_package_remote or default_remote
+
+                ret_upload_package = self._upload_package(pref,
                                                           index + 1, total, retry, retry_wait,
                                                           integrity_check,
-                                                          policy, upload_remote)
+                                                          policy, p_remote)
+
+                if (not cur_package_remote or not remote_name) and policy != UPLOAD_POLICY_SKIP:
+                    self._registry.prefs.set(pref, p_remote.name)
+
                 if ret_upload_package:
                     recorder.add_package(str(conan_ref), package_id)
 
-        if not defined_remote and policy != UPLOAD_POLICY_SKIP:
-            self._registry.refs.set(conan_ref, upload_remote.name)
-        
+        if (not cur_recipe_remote or not remote_name) and policy != UPLOAD_POLICY_SKIP:
+            self._registry.refs.set(conan_ref, recipe_remote.name)
```

Comment:
I would think that the registry should be updated just after uploading the recipe (line 103)

And also, I don't know why it is conditioned on ``not cur_recipe_remote or not remote_name`` (same comment as the one regarding packages).

Reply:
Agree about moving the registry update.
I think the condition can be simplified to `not cur_recipe_remote`, the same for packages. Let me check. The idea is: If something has a registry entry, then an upload won't change it.